src/type.js

```diff
@@ -0,0 +1,313 @@
+      if (!currentElement().disabled) {
+        const returnValue = await callback({eventOverrides})
+        Object.assign(eventOverrides, returnValue?.eventOverrides)
+      }
+    }
+  }
+
+  function calculateNewValue(newEntry) {
+    const {selectionStart, selectionEnd} = currentElement()
+    // can't use .maxLength property because of a jsdom bug:
+    // https://github.com/jsdom/jsdom/issues/2927
+    const maxLength = Number(currentElement().getAttribute('maxlength') ?? -1)
+    const value = currentValue()
+    let newValue
+    if (selectionStart === selectionEnd) {
+      if (selectionStart === 0) {
+        // at the beginning of the input
+        newValue = newEntry + value
+      } else if (selectionStart === value.length) {
+        // at the end of the input
+        newValue = value + newEntry
+      } else {
+        // in the middle of the input
+        newValue =
+          value.slice(0, selectionStart - 1) +
+          newEntry +
+          value.slice(selectionEnd)
+      }
+    } else {
+      // we have something selected
+      newValue =
+        value.slice(0, selectionStart) + newEntry + value.slice(selectionEnd)
+    }
+
+    if (maxLength < 0) {
+      return newValue
+    } else {
+      return newValue.slice(0, maxLength)
+    }
+  }
+
+  async function typeCharacter(char, {eventOverrides}) {
+    const key = char // TODO: check if this also valid for characters with diacritic markers e.g. úé etc
+    const keyCode = char.charCodeAt(0)
+
+    const keyDownDefaultNotPrevented = fireEvent.keyDown(currentElement(), {
+      key,
+      keyCode,
+      which: keyCode,
+      ...eventOverrides,
+    })
+
+    if (keyDownDefaultNotPrevented) {
+      await tick()
+
+      const keyPressDefaultNotPrevented = fireEvent.keyPress(currentElement(), {
+        key,
+        keyCode,
+        charCode: keyCode,
+        ...eventOverrides,
+      })
+
+      const newValue = calculateNewValue(key)
+
+      if (keyPressDefaultNotPrevented && newValue !== currentValue()) {
+        if (!currentElement().readOnly) {
+          await tick()
+
+          const {selectionStart} = currentElement()
+
+          fireEvent.input(currentElement(), {
+            target: {
+              value: newValue,
+            },
+          })
+
+          element.setSelectionRange?.(selectionStart + 1, selectionStart + 1)
+        }
+      }
+    }
+
+    await tick()
+
+    fireEvent.keyUp(currentElement(), {
+      key,
+      keyCode,
+      which: keyCode,
+      ...eventOverrides,
+    })
+  }
+
+  function modifier({name, key, keyCode, modifierProperty}) {
+    return {
+      [`{${name}}`]: ({eventOverrides}) => {
+        const newEventOverrides = {[modifierProperty]: true}
+
+        fireEvent.keyDown(currentElement(), {
+          key,
+          keyCode,
+          which: keyCode,
+          ...eventOverrides,
+          ...newEventOverrides,
+        })
+
+        return {eventOverrides: newEventOverrides}
```

Comment:
I think you want to return both `eventOverrides` AND `newEventOverrides` here. Otherwise multiple held modifiers don't work correctly (such as `userEvent.type('{ctrl}{meta}{enter}')`)

Reply:
Actually, we merge the old eventOverrides with the new ones elsewhere so we're good here 👍